File: functions/src/chat/system_prompt.test.ts
import {buildSystemPrompt} from "./system_prompt";

describe("buildSystemPrompt", () => {
  it("includes today's date", () => {
    const prompt = buildSystemPrompt({});
    const today = new Date().toISOString().split("T")[0];
    expect(prompt).toContain(`Today's date is ${today}`);
  });

  it("says no modules when empty", () => {
    const prompt = buildSystemPrompt({});
    expect(prompt).toContain("user has no modules yet");
  });

  it("includes module names and schemas", () => {
    const prompt = buildSystemPrompt({
      mod1: {
        name: "Expenses",
        schemas: {
          expense: {
            label: "Expense",
            fields: {
              amount: {type: "currency", label: "Amount", required: true},
              category: {
                type: "enumType",
                label: "Category",
                options: ["Food", "Transport"],
              },
            },
          },
        },
      },
    });

    expect(prompt).toContain("Expenses");
    expect(prompt).toContain("mod1");
    expect(prompt).toContain("amount (currency)");
    expect(prompt).toContain("[required]");
    expect(prompt).toContain("options=[Food, Transport]");
  });

  it("includes rule about only operating on existing modules", () => {
    const prompt = buildSystemPrompt({});
    expect(prompt).toContain("ONLY operate on modules the user already has");
  });

  it("includes database table names and CREATE TABLE SQL", () => {
    const prompt = buildSystemPrompt({
      hiking: {
        name: "Hiking Journal",
        description: "Plan hikes and log trail experiences",
        database: {
          tableNames: {hike: "m_hikes"},
          setup: [
            'CREATE TABLE IF NOT EXISTS "m_hikes" (' +
              "id TEXT PRIMARY KEY, " +
              "name TEXT NOT NULL, " +
              "distance REAL, " +
              "difficulty TEXT DEFAULT 'moderate'" +
              ")",
            'CREATE INDEX idx_hikes_name ON "m_hikes" (name)',
          ],
        },
      },
    });

    expect(prompt).toContain("Hiking Journal");
    expect(prompt).toContain("hiking");
    expect(prompt).toContain('Schema key "hike" → table "m_hikes"');
    expect(prompt).toContain("CREATE TABLE");
    expect(prompt).toContain("name TEXT NOT NULL");
    // Index statements should not be included
    expect(prompt).not.toContain("CREATE INDEX");
  });

  it("includes module settings when present", () => {
    const prompt = buildSystemPrompt({
      mod1: {
        name: "Budget",
        settings: {monthlyBudget: 2000, currency: "USD"},
      },
    });
    expect(prompt).toContain("monthlyBudget");
    expect(prompt).toContain("2000");
  });
});

File: test/core/ai/system_prompt_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:aj_assistant/core/ai/system_prompt.dart';
import 'package:aj_assistant/core/models/module.dart';
import 'package:aj_assistant/core/database/module_database.dart';

void main() {
  group('buildSystemPrompt', () {
    test('includes base instructions', () {
      final prompt = buildSystemPrompt([]);
      expect(prompt, contains('You are AJ'));
      expect(prompt, contains('personal assistant'));
    });

    test('includes today date', () {
      final prompt = buildSystemPrompt([]);
      expect(prompt, matches(RegExp(r'\d{4}-\d{2}-\d{2}')));
    });

    test('shows no modules message when empty', () {
      final prompt = buildSystemPrompt([]);
      expect(prompt, contains('no modules yet'));
    });

    test('includes module name and description', () {
      final module = Module(
        id: 'expenses',
        name: 'Expense Tracker',
        description: 'Track daily spending',
        database: const ModuleDatabase(
          tableNames: {'default': 'expenses_default'},
          setup: [
            'CREATE TABLE "expenses_default" (id TEXT PRIMARY KEY, amount REAL, category TEXT)',
          ],
        ),
      );
      final prompt = buildSystemPrompt([module]);
      expect(prompt, contains('Expense Tracker'));
      expect(prompt, contains('Track daily spending'));
      expect(prompt, contains('expenses_default'));
      expect(prompt, contains('CREATE TABLE'));
    });

    test('includes rules section', () {
      final prompt = buildSystemPrompt([]);
      expect(prompt, contains('RULES'));
      expect(prompt, contains('ONLY operate on modules'));
    });
  });
}
